### PR TITLE
feature: Add user-based OAuth2 login

### PR DESCRIPTION
**Pull request by alkihis**
**WIP**: This is not tested, as OAuth2 user-auth flow is restricted to a private beta for now.
All code is extracted from [this documentation](https://developer.twitter.com/en/docs/twitter-api/oauth2), all *could work* if the doc is exact.

Add the following public methods:
- `generateOAuth2AuthLink`: Non-async function that generate a authorize link with given scope.
- `loginWithOAuth2`: Complete the 3-legged OAuth flow with all the requirements
- `refreshOAuth2Token`: Same as login, but to obtain a refresh token. (OAuth2 user-tokens are limited in time)
- `revokeOAuth2Token`: To revoke an obtained token

New instance creation method, with `{ clientId: string }` (needed to generate OAuth2 user-flow related stuff).

### CODE REVIEW

**Comment by PLhery** [comment]
awesome work

File: src/client-mixins/request-param.helper.ts [hunk 1]
```diff
@@ -150,6 +154,37 @@ export class RequestParamHelpers {
     // Remove the query string
     return urlObject.href.slice(0, urlObject.href.length - urlObject.search.length);
   }
+
+  static applyRequestParametersToUrl(url: string, parameters: TRequestQuery) {
```
Comment (PLhery) [inline src/client-mixins/request-param.helper.ts:158]: [optional]
I'm not a fan of trying to parse the URL ourselves when we can leverage URL()

There are so many edge cases
 - ex. the ? could be in the # area (or in the auth area etc..) and therefore not be actually the start of the query string -
 
 I'm sure the code would be simpler like that
 
 ```
 const url= new URL('https://api.twitter.com/2/tweets/:id?query=string')
 url.pathname = url.pathname.replace(':id', 1234) // some more advanced magic there
return url.toString()

Comment (PLhery) [reply src/client-mixins/request-param.helper.ts:158]: Also, a quick description above would be nice :) (an example of what it does for instance), as the name is not crystal clear

**Review by PLhery — APPROVED** [review]

File: src/client-mixins/request-param.helper.ts (hunk 1 above)
Comment (alkihis) [reply src/client-mixins/request-param.helper.ts:158]: Yeah, it's a way more cleaned way :)
When I've built it, I thought new URL would crash before `:` is an invalid URL component, right? (but it seems to accomodate)

**Comment by deadcoder0904** [comment]
> * `loginWithOAuth2`: Complete the 3-legged OAuth flow with all the requirements

@alkihis Does this completely replace the need for https://github.com/PLhery/node-twitter-api-v2/blob/master/doc/auth.md?

I am currently doing it manually here → https://github.com/deadcoder0904/twitter-api-v2-3-legged-login-using-next-connect/ but would love to replace it if all I have to do is call `loginWithOAuth2` once.

I can't see any docs for it so not sure 🤔

**Comment by alkihis** [comment]
@deadcoder0904: absolutely not! It is the same flow as existing in OAuth 1.0a, but with OAuth 2 (authorize link > app acceptation > callback > get access token). 
In fact, this now flow is a little bit more restrictive, as OAuth 2 tokens now has an expiration date, thus it will be required to ask for a refreshed token after expiration of initial access token.

**For now, Twitter OAuth 2 user-flow is still in private beta**, it means you cannot use it in your projects without asking the permission to Twitter. It's why this flow isn't documented yet. Please do not use it (or complain if it doesn't work :)), at least for now.
This lib is ready for the future, but the future isn't now. 😄

**Comment by bedwardly-down** [comment]
Ah man. I was about to complain and do a table flip!!! 🤪

I’m excited for this though. I’m still new to oauth and whatnot but can’t wait to test this out. 😉

**Pull request merged**